AST: Stop diagnosing potentially unavailable declarations in unavailable contexts

tshortli opened the pull request:
Potential unavailability of a declaration has always been diagnosed in contexts that do not have a sufficient platform introduction constraint, even when those contexts are also unavailable on the target platform. This behavior is overly
strict, since the potential unavailability will never matter, but it's a longstanding quirk of availability checking. As a result, some source code has been written to work around this quirk by marking declarations as simultaneously unavailable and introduced for a given platform:

```
@available(macOS, unavailable, introduced: 15)
func unavailableAndIntroducedInMacOS15() {
  // ... allowed to call functions introduced in macOS 15.
}
```

When availability checking was refactored to be based on a constraint engine in https://github.com/swiftlang/swift/pull/79260, the compiler started effectively treating `@available(macOS, unavailable, introduced: 15)` as just `@available(macOS, unavailable)` because the introduction constraint was treated as lower priority and therefore superseded by the unavailability constraint. This caused a regression for the code that was written to work around the availability checker's strictness.

We could try to match the behavior from previous releases, but it's actually tricky to match the behavior well enough in the new availability checking architecture to fully fix source compatibility. Consequently, it seems like the best fix is actually to address this long standing issue and stop diagnosing potential unavailability in unavailable contexts. The main risk of this approach is source compatibility for regions of unavailable code. It's theoretically possible that restricting available declarations by introduction version in unavailable contexts is important to prevent ambiguities during overload resolution in some codebases. If we find that is a problem that is too prevalent, we may have to take a different approach.

Resolves rdar://147945883.

tshortli committed:
AST: Stop diagnosing potentially unavailable declarations in unavailable contexts.

Potential unavailability of a declaration has always been diagnosed in contexts
that do not have a sufficient platform introduction constraint, even when those
contexts are also unavailable on the target platform. This behavior is overly
strict, since the potential unavailability will never matter, but it's a
longstanding quirk of availability checking. As a result, some source code has
been written to work around this quirk by marking declarations as
simultaneously unavailable and introduced for a given platform:

```
@available(macOS, unavailable, introduced: 15)
func unavailableAndIntroducedInMacOS15() {
  // ... allowed to call functions introduced in macOS 15.
}
```

When availability checking was refactored to be based on a constraint engine in
https://github.com/swiftlang/swift/pull/79260, the compiler started effectively
treating `@available(macOS, unavailable, introduced: 15)` as just
`@available(macOS, unavailable)` because the introduction constraint was
treated as lower priority and therefore superseded by the unavailability
constraint. This caused a regression for the code that was written to work
around the availability checker's strictness.

We could try to match the behavior from previous releases, but it's actually
tricky to match the behavior well enough in the new availability checking
architecture to fully fix source compatibility. Consequently, it seems like the
best fix is actually to address this long standing issue and stop diagnosing
potential unavailability in unavailable contexts. The main risk of this
approach is source compatibility for regions of unavailable code. It's
theoretically possible that restricting available declarations by introduction
version in unavailable contexts is important to prevent ambiguities during
overload resolution in some codebases. If we find that is a problem that is too
prevalent, we may have to take a different approach.

Resolves rdar://147945883.

tshortli commented:
@swift-ci please test

tshortli commented:
@swift-ci please test

tshortli commented:
@swift-ci please test source compatibility

beccadax commented on test/Sema/availability_scopes.swift:
```diff
@@ -418,8 +418,23 @@ extension SomeEnum {
-func unavailableOnMacOSAndIntroduced() {
+@available(macOS, unavailable)
+@available(macOS, introduced: 52)
+ func unavailableOnMacOSAndIntroduced() {
+ }
```
Nit: Indentation went a little funky here.

beccadax reviewed (changes requested):
Since there's a (hopefully small) potential for source breaks here, please write this up in CHANGELOG.md.

Otherwise this LGTM, assuming source compatibility tests don't turn up any problems.

tshortli commented:
Source compatibility tests passed: https://ci.swift.org/job/swift-PR-source-compat-suite-macos/2375/

tshortli commented:
@swift-ci please smoke test

tshortli commented:
Thanks for the review @beccadax - I've updated `CHANGELOG.md` and also added a constraints test that demonstrates the new potential for ambiguity. Fortunately, platform introduction-based disambiguation has always been an unreliable technique (for clients that have a high enough deployment target, the reference will still be ambiguous) so I'm pretty confident this source break should be small.
